### PR TITLE
ENH: Translate `sph_harm` Cython->C++, add `sph_harm_all` gufunc

### PR DESCRIPTION
**Pull request by izaid**
This PR translates `sph_harm` from Cython to C++. It was reasonably straightforward.

As usual, I discovered some unexpected things going on. For instance, we still support passing `double` `m` and `n` which then get casted to integers. I maintained this behaviour.

Something else I found is `sph_harmonic_unsafe` in `_legacy.pxd`, see https://github.com/scipy/scipy/blob/main/scipy/special/_legacy.pxd#L39. This isn't public facing or part of any API, so can we just delete it?

cc @steppi

### CODE REVIEW

**Comment by izaid** [comment]
> Not related to this PR but I think we must address this folder and file naming scheme before it is too late and before it starts to require massive amount of work to iron out. I am already completely lost in terms of what is in which file just to review things. Let's do it after this PR if possible.

Totally agree on this. I think our general plan is to group functions by family (e.g. `legendre.h`, `bessel.h`, etc). The issue is as things get translated from the older libraries, they come in as a monolith of code and it's impossible for us to break them up in a single PR. So it's a bit of a process to get them into the right place.

**Comment by steppi** [comment]
> Not related to this PR but I think we must address this folder and file naming scheme before it is too late and before it starts to require massive amount of work to iron out. I am already completely lost in terms of what is in which file just to review things. Let's do it after this PR if possible.

I think two steps we can take are:

1. Come up with a good name for this C++ special function library.
2. Move it to `scipy/scipy/_lib` where submodules currently are, in preparation for it eventually becoming a submodule.

This will help us get away from the confusing `scipy/scipy/special/special` structure, and I think are good steps to take anyway.

**Comment by ilayn** [comment]
We can at least get the top level files in order, by what they do. Currently everything is `XXX_wrappers`,  `?special?.XXX` and `cython_XXX.pxd`. 

We can use things like 

`ufunc_definitions.XXX`,
`top_level_wrappers.XXX`

and so on, in other words call them by what they are doing. Also under special/special the header file names are identical to the library header only files `amos.h` -> `amos/amos.h`. I think we should use the word `special` very sparingly as it is name of everything in that module.

**Comment by steppi** [comment]
> I think we should use the word `special` very sparingly as it is name of everything in that module.

I think having a name for the C++ library will help with that, so that if the library is called `X`, things like`X_wrappers` will be clear. 

Also, I don’t think anyone likes things `amos.h`, `amos/amos.h`, it was just a temporary expedient.

Restructuring things sounds like a good next step. I completely agree that things are a bit of a mess now.

**Comment by izaid** [comment]
Right, I've added one more thing to this PR. A gufunc called `sph_harm_all` that returns a `(m + 1, n + 1, ...)` array, broadcasting over the `theta` and `phi` parameters. This didn't exist in SciPy before, but as you can see it is not so hard to do now that we have the gufunc machinery merged. And I had to make use of `mdspan` to compose functions and views at the C++ level. I think there are probably a lot of people who would like this.

More generally, a lot of the special functions in SciPy actually need to compute a sequence or table of values to return the result. Examples include the Legendre functions, the spherical harmonics, the Bessel functions, and so on. In many cases, what is wanted is the whole set of values, not just a single selection. For instance, this issue https://github.com/scipy/scipy/issues/19079#issuecomment-2045599609 has asked for exactly this for the Bessel functions. Sometimes, though, a selection is desired too.

We are now in a position where we can provide these things relatively easily by exposing them as generalised ufuncs. This PR only intends to do so for `sph_harm`. A question, however, is how these functions should be named. Already in SciPy we have `lpn`, `lpmn`, `pbdv_seq`, `obl_cv_seq`, and so on that compute the whole set **but without broadcasting**. (Well, `lpn`, and `lpmn` do with broadcasting as of the merge yesterday.) I'd propose that all the functions that compute a sequence with broadcasting be given a clear subscript to distinguish them from their non-set counterparts. The spherical harmonics are a good example, because the existing `sph_harm` does not compute a sequence and this PR adds in its counterpart `sph_harm_all` that does.

I've used the subscript `_all` in this PR. I'm happy to change it, but we should pick something and stick with it. And other functions like `lpmn` that do compute a set of values should be slowly adapted for consistency: like `sph_harm`, we'd want a `lpmn` that returns values for fixed `m` and `n` and a `lpmn_all` that returns the whole table of values. The existing `lpmn` currently returns the whole table, and the existing `sph_harm` currently returns a fixed selection, so it's inconsistent.

Other subscript options besides `_all` are `_seq`, `_table`, `_array`, `_multi`, and `_full`. I think `_all` makes the most sense here. `_seq` is too related to something 1D, and we already have examples where the result is not 1D. But open to suggestions!

**Comment by izaid** [comment]
Oh, and there is still work to do on this PR, I see some failing jobs, etc. Just wanted to start the discussion.

**Comment by ilayn** [comment]
> I think there are probably a lot of people who would like this.

I am one of them. Though we could also have done this `std::vector` but your C++-fu is more advanced than mine so not going to press on any points :) 

> The spherical harmonics are a good example, because the existing sph_harm does not compute a sequence and this PR adds in its counterpart sph_harm_all that does.

Can the current `sph_harm` be a special case of `sph_harm` conceptually without changing things too much in the signature? Then we can actually overhaul the functions themselves which would be amazing.

**Comment by izaid** [comment]
> Can the current `sph_harm` be a special case of `sph_harm` conceptually without changing things too much in the signature? Then we can actually overhaul the functions themselves which would be amazing.

Unfortunately, no, I don't think so. The current `sph_harm` takes arrays `m`, `n`, `theta`, and `phi` and returns output of shape `np.broadcast_shapes(m.shape, n.shape, theta.shape, phi.shape)`. The `sph_harm_all` written here takes integers `m` and `n` as well as arrays `theta` and `phi` and returns output of shape `(m + 1, n + 1) + np.broadcast_shapes(theta.shape, phi.shape)`. They are only the same if, for the existing `sph_harm`, `m` and `n` are integer arrays in order.

I think the behaviour of the existing `sph_harm` is useful and we want to retain it, in addition to the new `sph_harm_all`. What is not good is `lpmn` and a few other functions as they are now, which are really doing what these `_all` functions should be doing. `lpmn` probably needs to ultimately become `lpmn_all` with a new `lpmn` coming in that behaves like the current `sph_harm`.

**Comment by ilayn** [comment]
The shapes are not that important, the number of output args and their type is essential so that we don't break anything. Also `_all` means really not much in this context because all of what? is the natural question.

**Comment by izaid** [comment]
Right, I've updated this PR so `sph_harm_all` now returns the full `2 * m + 1` set of values, with the negative ones after the positive ones. This means the result can be indexed like `y[m, n]` where m can be positive or negative, and you get what you expect. I've seen this convention elsewhere, and it makes the whole thing very convenient.

**Comment by ilayn** [comment]
Just to be clear, what I was wondering is whether this `sph_harm_all` can be subsumed into `sph_harm` or not. So that we don't introduce a new public API function.

**Comment by izaid** [comment]
> Just to be clear, what I was wondering is whether this `sph_harm_all` can be subsumed into `sph_harm` or not. So that we don't introduce a new public API function.

It could be done by adding an appropriate keyword to `sph_harm`, if that's the way we want to go. In which case we are just toggling the behaviour. If we want to do this, we should probably think about it with the proposed [special extension](https://github.com/data-apis/array-api/issues/725) in mind.

**Comment by ilayn** [comment]
> If we want to do this, we should probably think about it with the proposed https://github.com/data-apis/array-api/issues/725 in mind.

Ah yes, good point. Then a single signature is even more important to keep the API surface small.

**Comment by steppi** [comment]
> Something else I found is `sph_harmonic_unsafe` in `_legacy.pxd`, see https://github.com/scipy/scipy/blob/main/scipy/special/_legacy.pxd#L39. This isn't public facing or part of any API, so can we just delete it?

OK, this actually is used. I even found a [comment](https://github.com/scipy/scipy/pull/15311#issuecomment-1003098339) where I referred to it, demonstrating that I was once aware of its purpose.

The wrappers in `_legacy.pxd` are used to print runtime warnings if someone inputs a `float` for an argument that should be an `int`. On main we see

```
In [2]: sph_harm(3.2, 5, [0.2, 0.4, 0.2], [0.1, 0.1, 0.1])
<ipython-input-2-78ce2280316f>:1: RuntimeWarning: floating point number truncated to an integer
  sph_harm(3.2, 5, [0.2, 0.4, 0.2], [0.1, 0.1, 0.1])
Out[2]: 
array([-0.00224728-0.00153745j, -0.00098665-0.00253782j,
       -0.00224728-0.00153745j])
```

but on `izaid:sph_harm` we have

```
In [5]: sph_harm(3.2, 5, [0.2, 0.4, 0.2], [0.1, 0.1, 0.1])
Out[5]: 
array([-0.00224728-0.00153745j, -0.00098665-0.00253782j,
       -0.00224728-0.00153745j])
```

I think we can't just drop the `RuntimeWarnings` here, so we need to find a way to preserve them. Eventually I think we should go through a deprecation cycle for non-integer arguments for all of these functions.

**Comment by izaid** [comment]
> RuntimeWarning: floating point number truncated to an integer

Right, but we just need that behaviour? I can make that happen in the extension module.

**Comment by steppi** [comment]
> > RuntimeWarning: floating point number truncated to an integer
> 
> Right, but we just need that behaviour? I can make that happen in the extension module.

Yes, you just need to raise the `RuntimeWarning` when floating point input is given. It doesn't matter how it's done.

**Comment by izaid** [comment]
> > > RuntimeWarning: floating point number truncated to an integer
> > 
> > 
> > Right, but we just need that behaviour? I can make that happen in the extension module.
> 
> Yes, you just need to raise the `RuntimeWarning` when floating point input is given. It doesn't matter how it's done.

Done. It looks identical to the old Cython, and warning raised.

Screenshot 2024-04-11 at 3 59 53 PM

**Comment by steppi** [comment]
@izaid, once the `RuntimeWarnings` are sorted out, I'd be happy to merge this if `sph_harm_all` is taken out of the public API (so that it can only be imported as

```python
 from scipy.special._gufuncs import _sph_harm_all
```

and it gets a docstring which says "Private function. This may be removed or modified at any time.".

File: scipy/special/_basic.py [hunk 1]
```diff
@@ -3433,3 +3434,21 @@ def zeta(x, q=None, out=None):
...
+
+    phi = np.asarray(phi)
+    if (not np.issubdtype(phi.dtype, np.inexact)):
+        phi = phi.astype(np.float64)
+
+    out = np.empty((2 * m + 1, n + 1) + np.broadcast_shapes(theta.shape, phi.shape),
+        dtype = np.result_type(1j, theta.dtype, phi.dtype))
+    __sph_harm_all(theta, phi, out = np.moveaxis(out, (0, 1), (-2, -1)))
+
+    return out
```
Comment (steppi) [inline scipy/special/_basic.py:3454]: `\n`

File: scipy/special/tests/test_sph_harm.py [hunk 2]
```diff
@@ -35,3 +36,21 @@ def Y11(theta, phi):
...
+    y_actual = _sph_harm_all(n_max, n_max, theta, phi)
+
+    for n in [0, 1, 2, 5, 10, 20, 50]:
+        for m in [0, 1, 2, 5, 10, 20, 50]:
+            if (m <= n):
+                y_desired = sc.sph_harm(m, n, theta, phi)
+                np.testing.assert_allclose(y_actual[m, n], y_desired, rtol = 1e-05)
+
+                y_desired = sc.sph_harm(-m, n, theta, phi)
+                np.testing.assert_allclose(y_actual[-m, n], y_desired, rtol = 1e-05)
```
Comment (steppi) [inline scipy/special/tests/test_sph_harm.py:56]: `\n`

File: scipy/special/_basic.py [hunk 3]
```diff
@@ -14,7 +14,8 @@
 from ._ufuncs import (mathieu_a, mathieu_b, iv, jv, gamma,
                       psi, hankel1, hankel2, yv, kv, poch, binom,
                       _stirling2_inexact)
-from ._gufuncs import _lpn, _lpmn, _clpmn, _lqn, _lqmn, _rctj, _rcty
+from ._gufuncs import (_lpn, _lpmn, _clpmn, _lqn, _lqmn, _rctj, _rcty,
+                       _sph_harm_all as __sph_harm_all)
```
Comment (steppi) [inline scipy/special/_basic.py:18]: How about `_sph_harm_all as _sph_harm_all_gufunc` to avoid these double leading underscores?

Comment (izaid) [reply scipy/special/_basic.py:18]: Sure.

**Review by steppi — APPROVED** [review]
I'm happy with this now and feel it's safe to merge since nothing has been added to the public API. Let's start a discussion for establishing a consistent convention for gufunc versions of existing functions like this.

**Pull request merged**